File: dotnet/Mcma.Worker/Worker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Threading.Tasks;
using Mcma.Core.Logging;

namespace Mcma.Worker
{
    internal class Worker : IWorker
    {
        internal Worker(IEnumerable<IWorkerOperationFilter> operations)
        {
            OperationHandlers = operations?.ToArray() ?? new IWorkerOperationFilter[0];
        }

        private IWorkerOperationFilter[] OperationHandlers { get; }

        public async Task DoWorkAsync(WorkerRequest requestContext)
        {
            var operationHandler = OperationHandlers.FirstOrDefault(oh => oh.Filter(requestContext))?.Handler;
            if (operationHandler == null)
                throw new Exception($"No handler found for '{requestContext.OperationName}' that can handle this request.");

            Logger.Debug("Handling worker operation '" + requestContext.OperationName + "' with handler of type '" + operationHandler.GetType().Name + "'");
            
            try
            {
                await operationHandler.ExecuteAsync(requestContext);
            }
            catch (Exception ex)
            {
                Logger.Error($"Failed to process worker operation '{requestContext.OperationName}'.");
                Logger.Exception(ex);
                
                throw;
            }
        }
    }
}

File: nodejs/mcma-worker/lib/worker.js
const { Logger } = require("mcma-core");
const WorkerRequest = require("./worker-request");

class Worker {
    constructor(operationFilters) {

        this.doWork = async (request) => {
            if (!(request instanceof WorkerRequest)) {
                throw new Error("request must be an instance of class WorkerRequest");
            }

            const operationFilter = operationFilters.find(oh => oh.filter(request));
            if (!operationFilter) {
                throw new Error(`No handler found for ${request.operationName} that can handle this request.`);
            }

            Logger.debug("Handling worker operation '" + request.operationName + "'...");

            try {
                await operationFilter.handler(request);
            } catch (e) {
                Logger.error(`Failed to process worker operation "${request.operationName}"`);
                Logger.exception(e);

                throw e;
            }
        };
    }
}

module.exports = Worker;